### PR TITLE
CI: 3.8 build

### LINKED ISSUE #26626: Setup python 3.8 builds
As the python 3.8 release date draws closer ( currently [scheduled](https://www.python.org/dev/peps/pep-0569/#release-schedule) for October this year ).

I was wondering if we want to set up python3.8 builds?

Given the 3.8 dev branch is available on [travis](https://docs.travis-ci.com/user/languages/python/) and shortly [azure](https://github.com/Microsoft/azure-pipelines-image-generation/issues/905).

Might be abit early for this - but it will mean we can catch any potential issues early + support python 3.8 upon release date.

Related issues: https://github.com/pandas-dev/pandas/issues/22714

I am happy to work on this - but would be good to get people's thoughts first.

Thanks!

### PR DESCRIPTION
**Pull request by jbrockmendel**
Travis config is a mystery to me, so this is trial-and-error until I get the hang of it.

Closes https://github.com/pandas-dev/pandas/issues/26626

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -30,27 +30,12 @@ matrix:
       - python: 3.5
 
     include:
-    - dist: trusty
+    - dist: bionic
```
Comment (WillAyd) [inline .travis.yml]: Slightly orthogonal but is there a reason to even specify this at all? Not sure why we need to deviate from whatever the default environment provides

**Review by WillAyd — COMMENTED** [review]
I think the failure is a result of changing the `dist`. This link might have the answer:

https://docs.travis-ci.com/user/trusty-to-xenial-migration-guide#1-services-support

I think would need to add that key for both mysql and psql

**Comment by jreback** [comment]
see #27266 as well here

**Comment by jbrockmendel** [comment]
For a moment I was amazed that the last "restore everything disabled during debugging" commit came back green.

Then I saw that I had mangled the Travis config so badly that Travis didn't run at all.

**Comment by TomAugspurger** [comment]
FYI, on the NumPy mailing list they announced that they built 3.8 wheels. Looks like https://github.com/MacPython/numpy-wheels/blob/master/.travis.yml 

Maybe https://github.com/MacPython/numpy-wheels/commit/9d8a0adf9ef8ba0c42c2221b5584a9460ffe1e50 + an update of multibuild in the pandas-wheels repo would do the trick.

**Comment by jbrockmendel** [comment]
> Maybe MacPython/numpy-wheels@9d8a0ad + an update of multibuild in the pandas-wheels repo would do the trick.

Does this replace the approach being taken in this PR? Or supplement?

**Comment by jbrockmendel** [comment]
@TomAugspurger this approach is now working.  I'll take a look at the MacPython/multibuild approach to see if its significantly cleaner, but we have this as a fallback.

File: ci/build38.sh [hunk 2]
```diff
@@ -0,0 +1,23 @@
+#!/bin/bash -e
+# Special build for python3.8 until numpy puts its own wheels up
+
+sudo apt-get install build-essential gcc xvfb
+pip install --no-deps -U pip wheel setuptools
+pip install python-dateutil pytz cython pytest pytest-xdist hypothesis
+
+git clone https://github.com/numpy/numpy
```
Comment (TomAugspurger) [inline ci/build38.sh]: In theory, you can replace this with a `pip install --pre -f https://7933911d6844c6c53a7d-47bd50c35cd79bd838daf386af554a83.ssl.cf2.rackcdn.com/ numpy`

**Review by TomAugspurger — COMMENTED** [review]
Glad to see we didn't need any code changes. That's reassuring.

**Comment by jbrockmendel** [comment]
Reverted to use git clone, with comment about possible pip alternative

File: .travis.yml [hunk 3]
```diff
@@ -71,24 +77,27 @@ before_install:
...
   - echo "sh -e /etc/init.d/xvfb start"
-  - sh -e /etc/init.d/xvfb start
+  - if [ "$JOB" != "3.8-dev" ]; then sh -e /etc/init.d/xvfb start; fi
   - sleep 3
 
 script:
   - echo "script start"
-  - source activate pandas-dev
+  - echo "$JOB"
+  - if [ "$JOB" != "3.8-dev" ]; then source activate pandas-dev; fi
```
Comment (jreback) [inline .travis.yml]: why is this not also called pandas-dev (the env name)?

File: ci/build38.sh [hunk 4]
```diff
@@ -0,0 +1,25 @@
+#!/bin/bash -e
+# Special build for python3.8 until numpy puts its own wheels up
+
```
Comment (jreback) [inline ci/build38.sh]: you should create a bare conda env here

**Review by jreback — CHANGES_REQUESTED** [review]

File: .travis.yml (hunk 3 above)
Comment (jbrockmendel) [reply .travis.yml]: You mean $JOB?  none of the other $JOB values have pandas-dev in the name

File: ci/build38.sh (hunk 4 above)
Comment (jbrockmendel) [reply ci/build38.sh]: how come?

Comment (jreback) [reply ci/build38.sh]: 3.8 should not be very different than a regular setup

Comment (TomAugspurger) [reply ci/build38.sh]: It has to be different until there are 3.8 packages (python especially, but also our dependencies).

I think Brock has about the minimal difference we can get right now.

**Comment by TomAugspurger** [comment]
@jreback thoughts on https://github.com/pandas-dev/pandas/pull/28730#discussion_r334575150?

3.8 is out now, so I'd like to get this in and 0.25.2 released as soon as possible (ideally by sometime tomorrow).

**Review by jreback — APPROVED** [review]

**Comment by TomAugspurger** [comment]
https://github.com/pandas-dev/pandas/issues/29001 for followup.

Thanks @jbrockmendel.

**Pull request merged**